### DISCUSSION TITLE
Optional Fetcher

### DISCUSSION THREAD

**Discussion by gitperson1980** [antonmedv/expr#291] [General]
In some circumstances with certain structs and as needed, would it be possible to create provisions to use fetcher?

**Comment by antonmedv** [comment]
Can you please explain a little bit more?

As vm.Fetcher was removed in latest Expr versions. It’s still possible to do the same logic using the ast.Patch functionality. 

I will write a doc on how to do it.

**Comment by gitperson1980** [comment]
```go
type Bar struct{
   D int
}

type Foo struct {
	A int
	B bool
	Mp map[string]*Bar   // has elements  MP1 MP2 MP3 MP4 MP5 MP6   MP1-3 is call heavily and would like to cache it as pointers as below
	mp1 *Bar   //lower case to avoid export in serialization but cannot be accessed by expr
	mp2 *Bar   //lower case to avoid export in serialization
	mp3 *Bar   //lower case to avoid export in serialization
}

FooExample = Foo{A: 1, B: true, Mp: make(map[string]*Bar)}
FooExample.Mp["MP1"]= Bar{}
FooExample.Mp["MP2"]= Bar{}
FooExample.Mp["MP3"]= Bar{}
FooExample.Mp["MP4"]= Bar{}
FooExample.Mp["MP5"]= Bar{}
FooExample.mp1=FooExample.Mp["MP1"]    //cache the most used
FooExample.mp2=FooExample.Mp["MP2"]    //cache the most used
FooExample.mp3=FooExample.Mp["MP3"]    //cache the most used

FooExample.mp1.D   // direct and avoids the overhead of a map lookup

**Comment by antonmedv** [comment]
> call heavily and would like to cache it as pointers as below

Expose MP1 and no map lookup will be done.

**Comment by gitperson1980** [comment]
I cannot expose it as I don't want MP1-3 added in a gob serialization routine in another process and increase the size of the serialized output with duplicate information. This is why it is lowercase. However it can be accessed by a fetcher. 

What might be nice is to be able to fetch if fetcher exists/defined  else fall though to standard fetch. 

```go
func (v *Foo) Fetch(i interface{}) interface{} {
	var string_i string
	if s, ok := i.(string); ok {
		string_i = s
	} else {
		string_i = fmt.Sprint(i)
	}
	switch string_i {
	case "mp1":
		return v.mp1
	case "mp2":
		return v.mp2
	case "mp3":
		return v.mp3
	}
	return nil
}

**Comment by antonmedv** [comment]
I see. What about losing a separate structure for Env?

**Comment by gitperson1980** [comment]
I don't understand the above suggestion. But if you mean use a mostly duplicate structure in Env .. then that would mean duplicated effort and size for the object that is very large in size and complexity. It would require a great deal of coordination to keep them in sync and introduce error and performance overhead. 

Adding an optional fetcher within the new optimized paradigm would help immensely to do some optimization tricks for the end user. And I don't mean just this example. Case/Switch is cheap especially if you have nested structs with their own fetchers. 

What would be nice is a combo fetch from fetcher when it is defined and a fall through to standard processing when it is not defined or when the fetcher returns nil. It would afford great latitude for the end user to optimize according to his needs.

**Comment by antonmedv** [comment]
Something like this:
```
env := Env {
    Map: &BigMap,
    MP1: BigMap["MP1"],
}
```
And use ast.Patcher to convert this
```
MP1 + MP100
```
 to this
```
MP1 + Fetch("MP100")
```

**Reply by gitperson1980** [reply]
There would be a lot of re-definitions into Env and I would have to do that every time I introduce a different Foo into the Env - which is often. That would cause a lot of map access every time I setup Env with a different Foo and negate the caching benefit which is done only once when it is done in Foo itself.

**Reply by antonmedv** [reply]
I did not get it. My suggested approach should be even faster than old Fetcher. 

If you don't what this optimization and what old Fetcher style behavior, use ast.Patch to replace all vars to some fetcher func:

```
Fetch("MP1") + Fetch("MP100")
```

**Reply by gitperson1980** [reply]
Not that your suggestions are bad... an optional fetcher provides quick and dirty mods to the end user where he chooses to deploy it prudently in concert with your recent optimizations. ast.Patch is brilliant and I do use it but is cumbersome to use en-masse with large amount of variables. It's a better option to use to enhance language constructs as a whole vs alter/intercept how large amount of variables are accessed.

**Reply by antonmedv** [reply]
So what do you suggest?

**Comment by gitperson1980** [comment]
I would benefit from a hybrid fetcher functionality as:

```go
func Fetch(from, i interface{}) interface{} {
  if fetcher, ok := from.(Fetcher); ok {
        value := fetcher.Fetch(i)
        if value != nil {
           return value
        }
  }
// from here
// fall through to standard processing
}
```

Same for:
```go 
func FetchField(from interface{}, field *Field) interface{} {

for i, x := range field.Path{
         // traverse path using fetcher(s) where fetchers are defined
         // else use standard offset index processing
        }
  }
}

```

**Reply by antonmedv** [reply]
Well, adding fetcher check, in my benchmarks, ads around 10% slowness.

**Reply by gitperson1980** [reply]
in v1.9.0 I had hacked the runtime to eliminate the " if fetcher, ok := from.(Fetcher); ok {", as below,  check and called fetcher  by default as all items had it defined - some had the default of nil where it was not needed. Is there a way to define an interface{} Fetcher() that returns nil when not implemented? This way no check is needed but you still get a nil rejection.

or use a runtime function parameter (useFetcher bool)

```go
if useFetcher {                                        // non Fetcher users do not pay penalty just a minor compare penalty.
if fetcher, ok := from.(Fetcher); ok {
      value:=from.(Fetcher).Fetch(i)
}else {
...

```

v1.9.0 old hack:

```go

rntime.go 

func fetch(from, i interface{}, nilsafe bool) interface{} {
   //if fetcher, ok := from.(Fetcher); ok {
      //value := fetcher.Fetch(i)

      value:=from.(Fetcher).Fetch(i)
      if value != nil {
         return value
      }
      if !nilsafe {
         panic(fmt.Sprintf("cannot fetch %v from %T", i, from))
      }
      return nil


```

**Reply by antonmedv** [reply]
One idea I have is to do this check on compile step. And add additional opcode: OpLoadFetch & OpFetchFetcher.

This was we can eliminate the check completely. Can you implement such op codes?

**Reply by gitperson1980** [reply]
> One idea I have is to do this check on compile step. And add additional opcode: OpLoadFetch & OpFetchFetcher.
> 
> This was we can eliminate the check completely. Can you implement such op codes?

Compile time check is brilliant.

I can certainly try to hack something together for your final approval. Can you distinguish the 2 with some explanation?
 OpLoadFetch & OpFetchFetcher.

**Reply by antonmedv** [reply]
They are really similar. OpLoad* are for loading from env vars. OpFetch* are for loading form top of the stack vm.pop().

**Reply by gitperson1980** [reply]
I will give it a go.

**Comment by gitperson1980** [comment]
runtime.go
```go 

type Fetcher interface {
	Fetch(interface{}) interface{}
}

func FetchFetcher(from, i interface{}) interface{} {
	value:=from.(Fetcher).Fetch(i)             //bareback,  as its been vetted to have a fetcher at compile time.
	if value != nil {
		return value
	}
	panic(fmt.Sprintf("cannot fetch %v from %T", i, from))
	//return nil
}
```
vm.go:

```go

		case OpFetchFetcher:
			b := vm.pop()
			a := vm.pop()
			vm.push(runtime.FetchFetcher(a, b))	

```

**Reply by gitperson1980** [reply]
I did not see an OpLoad* set of opcodes in opcodes.go.

**Reply by antonmedv** [reply]
https://github.com/antonmedv/expr/blob/master/vm/vm.go#L103

**Reply by gitperson1980** [reply]
Ok got it.

**Reply by gitperson1980** [reply]
I believe this should take care of the opFetch* opcdes. I will do the rest if you are ok with the followin.

vm.go
```go
		case OpFetch:
			b := vm.pop()
			a := vm.pop()
			vm.push(runtime.Fetch(a, b))

		case OpFetchFetcher:
			b := vm.pop()
			a := vm.pop()
			vm.push(runtime.FetchFetcher(a, b))
			
		case OpFetchField:
			a := vm.pop()
			vm.push(runtime.FetchField(a, program.Constants[arg].(*runtime.Field)))

		case OpFetchFieldFetcher:
			a := vm.pop()
			vm.push(runtime.FetchFieldFetcher(a, program.Constants[arg].(*runtime.Field)))
```

runtime.go

```go

type Fetcher interface {
	Fetch(interface{}) interface{}
}

func FetchFetcher(from, i interface{}) interface{} {
	value:=from.(Fetcher).Fetch(i)             //bareback,  as its been vetted to have a fetcher at compile time.
	if value != nil {
		return value
	}
	panic(fmt.Sprintf("cannot fetch %v from %T", i, from))
	//return nil
}

func FetchFieldFetcher(from interface{},  field *Field) interface{} {
	value:=from
	for _, nextPathItem := range field.Path {
		// assumption that the path/field chain all has been checked for fetchers at compile time.
		// otherwise path chain has to be broken up between ones with fetchers submitted to this function
		// and ones without submitted to the regular FetchField function
		value=value.(Fetcher).Fetch(nextPathItem)
		if value==nil {
			panic(fmt.Sprintf("cannot fetch %v from %T", nextPathItem, value))
		}
	}
	return value
}

```

**Reply by antonmedv** [reply]
Let's ditch FetchFieldFetcher, and make it as a break point. No neet to do this optimization here.

**Reply by gitperson1980** [reply]
> Let's ditch FetchFieldFetcher, and make it as a break point. No neet to do this optimization here.

Sorry for the delay I had some things to attend to on Sunday. However, I was a bit confused about the "break point" part. Do you mean you would like to stick with FetchFetcher for everything with a fetcher and reserve FieldFetcher type access for only chain of paths without fetcher and hence abandon optimization there? I assume then fetcher will be called the old way as runtime traverses down a struct for a chain of paths without fetchers.

**Reply by mdmcconnell** [reply]
Hhi, I'm curious - what was the eventual result of this discussion?  I believe I'm trying to do something analogous, where the value of one  type of struct should always be read by calling a certain method.  Creating a near copy of env was one possibility, but messy as pointed out above; I've also tried patching the tree, and not understood properly how to do that.  This might be a third option - is the FetchFetcher functionality implemented?  How to use it?

**Reply by antonmedv** [reply]
It’s not implemented currently. One way to do, it is the aperture which should be superior to the future interface.

**Reply by mdmcconnell** [reply]
I'm afraid I did not understand - what is the aperture?  I do not remember seeing this concept in expr docs or code.

**Reply by antonmedv** [reply]
Yeah, looks like my iPhone stations get us run

**Reply by mdmcconnell** [reply]
Looks like possibly an auto-complete issue here.  I will maintain approach of trying to patch the tree with a CallNode on nodes that implement the certain interface.  But what are the required members for the CallNode and its Callee MemberNode?  With only ast.CallNode{ Callee: &ast.MemberNode{Node: *node, Name: "Method"}, I am getting "panic: undefined node type (<nil>)".  Same as https://go.dev/play/p/wkZtzPYBKF-

**Reply by mdmcconnell** [reply]
Eventually answered my own question in https://github.com/antonmedv/expr/discussions/368
